crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -140,10 +140,18 @@ where
     DB: Database,
 {
     fn header(&self, block_hash: &BlockHash) -> ProviderResult<Option<Header>> {
+        if let Some(block_state) = self.canonical_in_memory_state.state_by_hash(*block_hash) {
+            return Ok(Some(block_state.block().block().header.header().clone()));
```

Comment:
doing mem lookups first is the right order because we only remove from mem after we flushed